### ISSUE TITLE
Conversion to GTF does not convert phase to frame

### ISSUE THREAD

**Issue by channsoden** [gpertea/gffread#29]
Hello,

My understanding is that the GFF version 3 gives the phase in 8th field, while GTF gives the frame in the 8th field, and that these are not equivalent. See [the Wikipedia on GFF](https://en.wikipedia.org/wiki/General_feature_format). As such, frame = (3 - phase) % 3, and phase = (3 - frame) % 3.

The output to GTF (-T) does not seem to change this field, resulting in erroneous and somewhat nonsensical entries. For example, this GFF:

> ##gff-version 3
> scaffold1.1	AUGUSTUS	gene	26663	27325	0.99	+	.	ID=g6
> scaffold1.1	AUGUSTUS	transcript	26663	27325	0.99	+	.	ID=g6.t1;Parent=g6
> scaffold1.1	AUGUSTUS	start_codon	26663	26665	.	+	0	Parent=g6.t1
> scaffold1.1	AUGUSTUS	intron	26898	27016	1	+	.	Parent=g6.t1
> scaffold1.1	AUGUSTUS	intron	27089	27191	0.99	+	.	Parent=g6.t1
> scaffold1.1	AUGUSTUS	CDS	26663	26897	1	+	0	ID=g6.t1.cds;Parent=g6.t1
> scaffold1.1	AUGUSTUS	CDS	27017	27088	1	+	2	ID=g6.t1.cds;Parent=g6.t1
> scaffold1.1	AUGUSTUS	CDS	27192	27322	0.99	+	2	ID=g6.t1.cds;Parent=g6.t1
> scaffold1.1	AUGUSTUS	stop_codon	27323	27325	.	+	0	Parent=g6.t1

Produces this GTF:

> scaffold1.1	AUGUSTUS	CDS	26663	26897	1.00	+	0	transcript_id "g6.t1"; gene_id "g6";
> scaffold1.1	AUGUSTUS	CDS	27017	27088	1.00	+	2	transcript_id "g6.t1"; gene_id "g6";
> scaffold1.1	AUGUSTUS	CDS	27192	27325	0.99	+	2	transcript_id "g6.t1"; gene_id "g6";

But it should be:

> scaffold1.1	AUGUSTUS	CDS	26663	26897	1.00	+	0	transcript_id "g6.t1"; gene_id "g6";
> scaffold1.1	AUGUSTUS	CDS	27017	27088	1.00	+	1	transcript_id "g6.t1"; gene_id "g6";
> scaffold1.1	AUGUSTUS	CDS	27192	27325	0.99	+	1	transcript_id "g6.t1"; gene_id "g6";

In this case the protein sequence is still obvious, since the concatenated CDS is identical. However, my annotation has some CDS in the following perplexing format:

> scaffold22.1	AUGUSTUS	CDS	279	1165	0.21	+	2	ID=g5177.t1.cds;Parent=g5177.t1
> scaffold22.1	AUGUSTUS	stop_codon	1166	1168	.	+	0	Parent=g5177.t1

This CDS is clearly 2 bases too long, but taking account the phase it is still intelligible. The corresponding GTF entry, though, would be frame-shifted:

> scaffold22.1	AUGUSTUS	CDS	279	1168	0.21	+	2	transcript_id "g5177.t1"; gene_id "g5177";

Best,
Christopher

**Comment by gpertea** [comment]
Indeed I haven't touched that field because I sincerely believe it to have the same meaning in both GFF3 and GTF: that of "phase". Not sure how/when that section in the Wikipedia article came around to needlessly complicate things even further for this ancient format (which should be obsolete but I know it isn't), and in particular for this rather useless field.. (imo it was unnecessary to dedicate a whole column just for this rarely needed information, since only partial CDS features really make use of it so it should've been just stored as an attribute for the (1st?) CDS feature)
I've been handling GTF for many years (even those produced by some gene finders) and never encountered such a distinction for that field, I've always treated it as "phase". That's because the GTF specification/documentation I have been using all this time was based on this page: 
http://mblab.wustl.edu/GTF22.html
It's true that they call it "frame" there but the description really sounds the same with "phase" to me, i.e. how many bases to "remove" from that CDS region in order to get to the next full codon. Have I misunderstood that all this time? (it's possible.. truth be told never cared much about that "phase" field as you can tell).
I am now very curious what was the reference GTF specification document that the Wikipedia article author used in order to claim such a distinction for that field.. I see the the GFF3 format specification clearly documented and referenced, but no good GTF reference documentation to support that distinction (I myself only found the one linked above).
My GTF/GFF3 knowledge, linking to the GTF2 and GFF3 reference documentation I've been using, was summarized for years in this web page: http://ccb.jhu.edu/software/stringtie/gff.shtml)

**Comment by channsoden** [comment]
I agree, the GTF2 specification has the same description for "frame" as GFF3 does for "phase", so it looks like they are indeed equivalent and no change needs be made. It looks like my previous understanding and the Wikipedia is wrong.

A notation system where the "frame" represents the position in the codon of the first base seems equally rational, although that is clearly not what is used here. I wonder, though, if the distinction is clear for all the authors of tools that output GFF? It makes me worry that the field may be entirely arbitrary in practice.

I second your opinions about GTF being obsolete and the frame/phase column being unhelpful.

**Comment by channsoden** [comment]
Looks like the Wikipedia page has been like that since March 2012 ¯\_(ツ)_/¯
